dvc/remote/base.py

```diff
@@ -497,22 +508,43 @@ def _cache_is_copy(self, path_info):
         self.cache_type_confirmed = True
         return self.cache_types[0] == "copy"
 
-    def _save_dir(self, path_info, checksum, save_link=True):
-        cache_info = self.checksum_to_path_info(checksum)
-        dir_info = self.get_dir_cache(checksum)
+    def _save_dir(self, path_info, checksum, save_link=True, tree=None):
+        if tree:
+            checksum = self._save_tree(path_info, tree)
+        else:
+            dir_info = self.get_dir_cache(checksum)
```

Comment:
We've discussed this before, but just to clarify: I suppose we need this because we are not ready to make `_collect_dir` use the tree, just yet, right? Because it uses `self.walk_files`, which raises another question of abstracting tree-like Remote methods somehow (maybe in RemoteTree's or something)?

Reply:
So we can add tree support in `_collect_dir`, but it made the workflow look a bit strange to me, since `_collect_dir` is used in `get_dir_checksum`. If you go off of the current behavior for `fetch`ing external git files, we would do something along the lines of:
```
# get a dir checksum for the git dir via `get_dir_checksum()`/`_collect_dir()`/`tree.walk_files()`
dir_cache_info = cache.save_info(path, tree=tree)
# walk the tree again and save git file objects from tree
cache.save(path, dir_cache_info, tree=tree)
```

But walking the tree twice like this seems unnecessary for this use case, since we can save objects from the tree and collect the dir cache info at the same time during a single walk. I wasn't sure if saving objects from tree during the walk would be desired behavior for `_collect_dir` (because saving objects in a `get_dir_checksum` call definitely seems like non-desired behavior).

Thinking about it now, what we could do is add the `tree` param to `_collect_dir`, and if it's called with a tree then we just save the objects from tree during `walk_files`, and use `_collect_dir` here in place of `_save_tree`. And when `_collect_dir` is used w/o a tree param, it doesn't save any objects during `walk_files` and behaves the same way as before for `get_dir_checksum` calls.